src/Twig/Extension/SonataAdminExtension.php

```diff
@@ -521,162 +571,149 @@ public function getUrlSafeIdentifier($model, ?AdminInterface $admin = null)
-                            $text = $this->translator->trans($text, [], $catalogue);
-                        // NEXT_MAJOR: Remove this check
-                        } elseif (method_exists($fieldDescription->getAdmin(), 'trans')) {
-                            $text = $fieldDescription->getAdmin()->trans($text, [], $catalogue);
-                        }
-                    }
-
-                    $xEditableChoices[] = [
-                        'value' => $value,
-                        'text' => $text,
-                    ];
-                }
-            }
+        if ('sonata_deprecation_mute' !== (\func_get_args()[1] ?? null)) {
+            @trigger_error(sprintf(
+                'The %s method is deprecated in favor of XEditableExtension::getXEditableChoices since version 3.x and will be removed in 4.0.',
+                __METHOD__
+            ), E_USER_DEPRECATED);
         }
 
-        if (false === $fieldDescription->getOption('required', true)
-            && false === $fieldDescription->getOption('multiple', false)
-        ) {
-            $xEditableChoices = array_merge([[
-                'value' => '',
-                'text' => '',
-            ]], $xEditableChoices);
+        if (null === $this->xEditableExtension) {
+            $this->xEditableExtension = new XEditableExtension($this->translator, $this->xEditableTypeMapping);
         }
 
-        return $xEditableChoices;
+        return $this->xEditableExtension->getXEditableChoices($fieldDescription);
     }
 
     /**
+     * NEXT_MAJOR: Remove this method.
+     *
      * Returns a canonicalized locale for "moment" NPM library,
      * or `null` if the locale's language is "en", which doesn't require localization.
      *
      * @return string|null
+     *
+     * @deprecated since sonata-project/admin-bundle 3.x and will be removed in 4.0
      */
     final public function getCanonicalizedLocaleForMoment(array $context)
     {
-        $locale = strtolower(str_replace('_', '-', $context['app']->getRequest()->getLocale()));
-
-        // "en" language doesn't require localization.
-        if (('en' === $lang = substr($locale, 0, 2)) && !\in_array($locale, ['en-au', 'en-ca', 'en-gb', 'en-ie', 'en-nz'], true)) {
-            return null;
+        if ('sonata_deprecation_mute' !== (\func_get_args()[1] ?? null)) {
+            @trigger_error(sprintf(
+                'The %s method is deprecated in favor of CanonicalizeExtension::getCanonicalizedLocaleForMoment since version 3.x and will be removed in 4.0.',
+                __METHOD__
+            ), E_USER_DEPRECATED);
         }
 
-        foreach (self::MOMENT_UNSUPPORTED_LOCALES as $language => $locales) {
-            if ($language === $lang && !\in_array($locale, $locales, true)) {
-                $locale = $language;
-            }
+        if (null === $this->canonicalizeExtension) {
+            $this->canonicalizeExtension = new CanonicalizeExtension();
```

Comment:
```suggestion
            $requestStack = new RequestStack();
            $requestStack->push($context['app']->getRequest());
            $this->canonicalizeExtension = new CanonicalizeExtension($requestStack);
```
I guess you can do something like this if we don't inject `request_stack` here, the same for the other call to `canonicalizeExtension`.

Reply:
Thanks. I overlooked that $context contains request which can be pushed into requestStack